cleverhans/attacks.py

```diff
@@ -1,4 +1,351 @@
+from abc import ABCMeta
 import numpy as np
+import warnings
+
+from .utils import random_targets
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, pred, back='tf', sess=None, params={}):
+        """
+        :param pred: The model's symbolic output.
+        :param back: The backend to use. Either 'tf' (default) or 'th'.
+        :param params: Parameter dictionary used by child classes.
+        """
+        if not back == 'tf' or back == 'th':
+            raise Exception("Backend argument must either be 'tf' or 'th'.")
+        if back == 'tf' and sess is None:
+            raise Exception("A tf session was not provided in sess argument.")
+        if back == 'th' and sess is not None:
+            raise Exception("A session should not be provided when using th.")
+
+        # Prepare attributes
+        self.x = None
+        self.pred = pred
+        self.back = back
+        self.sess = sess
+        self.nb_calls_generate = 0
+        self.default_graph = None
+
+    def generate(self, x):
+        """
+        Generate the attack's symbolic graph for adversarial examples. This
+        method should be overriden in any child class that implements an
+        attack that is expressable symbolically.
+        :param x: The model's symbolic inputs.
+        :return: A symbolic representation of the adversarial examples.
+        """
+        # Keep track of the number of calls to warn when more than one default
+        # graph to run if generating adversarial examples numerically
+        self.nb_calls_generate += 1
+
+        self.x = x
+
+        if self.back == 'tf':
+            import tensorflow as tf
+            wrapper = tf.py_func(self.generate_np, [self.x], tf.float32)
+            if self.nb_calls_generate == 1:
+                self.default_graph = wrapper
+            return wrapper
+        else:
+            # TODO()
+            return False
+
+    def generate_np(self, X, params={'batch_size': 128}):
```

Comment:
For `generate`, there shouldn't be any requirement that the inputs are placeholders. Any symbolic tensor should work.

If `generate` describes the symbolic graph for a single minibatch, then I would expect `generate_np` to make the adversarial examples for just one minibatch. I wouldn't expect it to loop over all the examples in the training set.

Also, in practice, if you're generating adversarial examples for an entire dataset, you're probably doing something wrong. You probably want to do something like compute accuracy on adversarial examples for the whole dataset, which should not require holding an adversarial version of the whole dataset in memory at the same time.